sphinx/util/logging.py

```diff
@@ -8,9 +8,293 @@
+
+if False:
+    # For type annotation
+    from typing import Any, Generator, IO, Tuple  # NOQA
+    from docutils import nodes  # NOQA
+    from sphinx.application import Sphinx  # NOQA
+
+
+VERBOSE = 15
+DEBUG2 = 5
+
+LEVEL_NAMES = defaultdict(lambda: logging.WARNING)  # type: Dict[str, int]
+LEVEL_NAMES.update({
+    'CRITICAL': logging.CRITICAL,
+    'SEVERE': logging.CRITICAL,
+    'ERROR': logging.ERROR,
+    'WARNING': logging.WARNING,
+    'INFO': logging.INFO,
+    'VERBOSE': VERBOSE,
+    'DEBUG': logging.DEBUG,
+    'DEBUG2': DEBUG2,
+})
+
+VERBOSITY_MAP = defaultdict(lambda: 0)  # type: Dict[int, int]
+VERBOSITY_MAP.update({
+    0: logging.INFO,
+    1: VERBOSE,
+    2: logging.DEBUG,
+    3: DEBUG2,
+})
+
+COLOR_MAP = defaultdict(lambda text: text)  # type: Dict[int, unicode]
+COLOR_MAP.update({
+    logging.WARNING: 'darkred',
+    logging.DEBUG: 'darkgray',
+    DEBUG2: 'lightgray',
+})
+
+
+def getLogger(name):
+    # type: (str) -> SphinxLoggerAdapter
+    """Get logger wrapped by SphinxLoggerAdapter."""
+    return SphinxLoggerAdapter(logging.getLogger(name), {})
+
+
+def convert_serializable(records):
+    """Convert LogRecord serializable."""
+    for r in records:
+        # extract arguments to a message and clear them
+        r.msg = r.getMessage()
+        r.args = ()
+
+
+class SphinxWarningLogRecord(logging.LogRecord):
+    """Log record class supporting location"""
+    location = None  # type: Any
+
+    def getMessage(self):
+        # type: () -> str
+        message = super(SphinxWarningLogRecord, self).getMessage()
+        location = getattr(self, 'location', None)
+        if location:
+            message = '%s: WARNING: %s' % (location, message)
+        elif 'WARNING:' not in message:
+            message = 'WARNING: %s' % message
+
+        return message
+
+
+class SphinxLoggerAdapter(logging.LoggerAdapter):
+    """LoggerAdapter allowing ``type`` and ``subtype`` keywords."""
+
+    def warn_node(self, message, node, **kwargs):
+        # type: (unicode, nodes.Node, Any) -> None
+        """Emit a warning for specific node.
+
+        :param message: a message of warning
+        :param node: a node related with the warning
+        """
+        (source, line) = get_source_line(node)
+        if source and line:
+            kwargs['location'] = "%s:%s" % (source, line)
+        elif source:
+            kwargs['location'] = "%s:" % source
+        elif line:
+            kwargs['location'] = "<unknown>:%s" % line
+
+        self.warning(message, **kwargs)
+
+    def log(self, level, msg, *args, **kwargs):
+        # type: (Union[int, str], unicode, Any, Any) -> None
+        if isinstance(level, int):
+            super(SphinxLoggerAdapter, self).log(level, msg, *args, **kwargs)
+        else:
+            levelno = LEVEL_NAMES.get(level)
+            super(SphinxLoggerAdapter, self).log(levelno, msg, *args, **kwargs)
+
+    def verbose(self, msg, *args, **kwargs):
```

Comment:
Is this really needed?
This is used only in `Sphinx.setup_extension()` and `sphinx.ext.viewcode`.